sphinx/builders/epub3.py

```diff
@@ -223,31 +118,33 @@ def build_navlist(self, navnodes):
                 continue
             if node['level'] > self.config.epub_tocdepth:
                 continue
-            usenodes.append(node)
-        for i, node in enumerate(usenodes):
-            curlevel = node['level']
-            if curlevel == level + 1:
-                navlist.append(self.begin_navlist_block(level))
-            while curlevel < level:
-                level -= 1
-                navlist.append(self.end_navlist_block(level))
-            level = curlevel
-            if i != len(usenodes) - 1 and usenodes[i + 1]['level'] > level:
-                has_child = True
+
+            navpoint = NavPoint(node['text'], node['refuri'], [])
+            if node['level'] == level:
+                navstack.pop()
+                navstack[-1].children.append(navpoint)
+                navstack.append(navpoint)
+            elif node['level'] == level + 1:
+                level += 1
+                navstack[-1].children.append(navpoint)
+                navstack.append(navpoint)
+            elif node['level'] < level:
+                while node['level'] < len(navstack):
+                    navstack.pop()
+                level = node['level']
+                navstack[-1].children.append(navpoint)
+                navstack.append(navpoint)
             else:
-                has_child = False
-            navlist.append(self.new_navlist(node, level, has_child))
-        while level != 1:
-            level -= 1
-            navlist.append(self.end_navlist_block(level))
-        return '\n'.join(navlist)
+                raise
```

Comment:
@tk0miya I happened to realize it, I think this cause `RuntimeError: No active exception to reraise`

Reply:
Oops. This is py2 style... I'll fix this soon.

Note: I suppose to never reach here.